File: OpenWA/sdk/python/openwa/resources/status.py
"""Status (Stories) resource — WhatsApp status updates.

Backed by ``src/modules/status/status.controller.ts``.
NOTE: this is WhatsApp "Status/Stories", distinct from session lifecycle status.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._http import quote_segment
from ..types import (
    SendImageStatusRequest,
    SendTextStatusRequest,
    SendVideoStatusRequest,
    SendVoiceStatusRequest,
    StatusMedia,
    StatusRecord,
    StatusResult,
)

if TYPE_CHECKING:
    from .._http import HttpExecutor


class StatusResource:
    def __init__(self, http: "HttpExecutor") -> None:
        self._http = http

    def list(self, session_id: str) -> dict[str, list[StatusRecord]]:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/status")

    def from_contact(self, session_id: str, contact_id: str) -> dict[str, list[StatusRecord]]:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/status/{quote_segment(contact_id)}")

    def media(self, session_id: str, status_id: str) -> StatusMedia:
        """Fetch the stored media bytes for a status update (404 when there is no stored media)."""
        data, content_type = self._http.request_bytes(
            "GET", f"/api/sessions/{quote_segment(session_id)}/status/{quote_segment(status_id)}/media"
        )
        return {"data": data, "contentType": content_type}

    def send_text(self, session_id: str, body: SendTextStatusRequest) -> StatusResult:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/status/send-text", body=body)

    def send_image(self, session_id: str, body: SendImageStatusRequest) -> StatusResult:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/status/send-image", body=body)

    def send_video(self, session_id: str, body: SendVideoStatusRequest) -> StatusResult:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/status/send-video", body=body)

    def send_voice(self, session_id: str, body: SendVoiceStatusRequest) -> StatusResult:
        """Post an audio status as a voice note (OPERATOR).

        WhatsApp plays a status voice note only as Ogg/Opus and neither engine transcodes, so
        convert first with ``media.convert_voice`` and post the base64 it returns.
        """
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/status/send-voice", body=body)

    def delete(self, session_id: str, status_id: str) -> None:
        self._http.request("DELETE", f"/api/sessions/{quote_segment(session_id)}/status/{quote_segment(status_id)}")

File: OpenWA/sdk/javascript/src/resources/status.ts
/**
 * Status (Stories) resource — WhatsApp status updates.
 *
 * Backed by `src/modules/status/status.controller.ts`.
 * NOTE: this is WhatsApp "Status/Stories", distinct from session lifecycle status.
 * @packageDocumentation
 */

import { encodeSegment } from '../http.js';
import type { BinaryResponse } from '../http.js';
import type { OpenWAClient } from '../client.js';
import type {
  SendImageStatusRequest,
  SendTextStatusRequest,
  SendVideoStatusRequest,
  SendVoiceStatusRequest,
  StatusRecord,
  StatusResult,
} from '../types.js';

export class StatusResource {
  constructor(private readonly client: OpenWAClient) {}

  /** Get all status updates. */
  list(sessionId: string): Promise<{ statuses: StatusRecord[] }> {
    return this.client.request<{ statuses: StatusRecord[] }>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/status`,
    });
  }

  /** Get status updates from a specific contact. */
  fromContact(sessionId: string, contactId: string): Promise<{ statuses: StatusRecord[] }> {
    return this.client.request<{ statuses: StatusRecord[] }>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/${encodeSegment(contactId)}`,
    });
  }

  /** Fetch the stored media bytes for a status update (404 when there is no stored media). */
  media(sessionId: string, statusId: string): Promise<BinaryResponse> {
    return this.client.requestBytes({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/${encodeSegment(statusId)}/media`,
    });
  }

  /** Post a text status update. */
  sendText(sessionId: string, body: SendTextStatusRequest): Promise<StatusResult> {
    return this.client.request<StatusResult>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/send-text`,
      body,
    });
  }

  /** Post an image status update. */
  sendImage(sessionId: string, body: SendImageStatusRequest): Promise<StatusResult> {
    return this.client.request<StatusResult>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/send-image`,
      body,
    });
  }

  /** Post a video status update. */
  sendVideo(sessionId: string, body: SendVideoStatusRequest): Promise<StatusResult> {
    return this.client.request<StatusResult>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/send-video`,
      body,
    });
  }

  /**
   * Post an audio status as a voice note. No caption: WhatsApp has nowhere to render one.
   *
   * WhatsApp plays a status voice note only as Ogg/Opus and neither engine transcodes, so convert
   * first with `media.convertVoice` and post the base64 it returns. Requires an OPERATOR key.
   */
  sendVoice(sessionId: string, body: SendVoiceStatusRequest): Promise<StatusResult> {
    return this.client.request<StatusResult>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/send-voice`,
      body,
    });
  }

  /** Delete a status update by id. */
  delete(sessionId: string, statusId: string): Promise<void> {
    return this.client.request<void>({
      method: 'DELETE',
      path: `/api/sessions/${encodeSegment(sessionId)}/status/${encodeSegment(statusId)}`,
    });
  }
}
